extras/module-node/duk_module_node.c

```diff
@@ -0,0 +1,248 @@
+/*
+ * Node.js-like require() and module loader
+ *
+ * https://nodejs.org/api/modules.html
+ */
+
+#include "duktape.h"
+#include "duk_module_node.h"
+
+static duk_int_t duk__eval_module_helper (duk_context *ctx);
+static void duk__push_module_object (duk_context *ctx, const char *id);
+
+static duk_resolve_module_function resolve_module_cb;
+static duk_load_module_function load_module_cb;
+
+static duk_bool_t duk__get_cached_module(duk_context *ctx, const char *id) {
+	duk_push_global_stash(ctx);
+	duk_get_prop_string(ctx, -1, "\xffModuleCache");
+	if (duk_get_prop_string(ctx, -1, id)) {
+		duk_remove(ctx, -2);
+		duk_remove(ctx, -2);
+		return 1;
+	}
+	else {
+		duk_pop_3(ctx);
+		return 0;
+	}
+}
+
+/*
+ * NOTE: duk__put_cached_module() leaves the module object on the
+ * stack as a convenience to the caller.
+ */
+static void duk__put_cached_module(duk_context *ctx) {
+	/* ... module */
+
+	duk_push_global_stash(ctx);
+	duk_get_prop_string(ctx, -1, "\xffModuleCache");
+	duk_dup(ctx, -3);
+
+	/* ... module g_stash mod_cache module */
+
+	duk_get_prop_string(ctx, -1, "id");
+	duk_dup(ctx, -2);
+	duk_put_prop(ctx, -4);
+
+	duk_pop_3(ctx);  /* ... module */
+}
+
+static void duk__del_cached_module(duk_context *ctx, const char *id) {
+	duk_push_global_stash(ctx);
+	duk_get_prop_string(ctx, -1, "\xffModuleCache");
+	duk_del_prop_string(ctx, -1, id);
+	duk_pop_2(ctx);
+}
+
+static duk_ret_t duk__handle_require(duk_context *ctx) {
+	/*
+	 *  Value stack handling in this function is a bit sloppy but should be
+	 *  correct.  Since this is a Duktape/C call, we're relying on Duktape to
+	 *  clean up the extra stack garbage for us.
+	 */
+
+	const char *id;
+	duk_ret_t nrets;
+	const char *parent_id = NULL;
+	duk_idx_t module_idx;
+
+	duk_push_current_function(ctx);
+	duk_get_prop_string(ctx, -1, "\xffModuleID");
+	parent_id = duk_get_string(ctx, -1);
+
+	/* id require parent_id */
+
+	id = duk_require_string(ctx, 0);
+
+	duk_push_string(ctx, parent_id);
+	duk_push_string(ctx, id);
+	nrets = resolve_module_cb(ctx);
+	if (nrets > 1) {
+		/* Pop off extraneous return values. */
+		duk_pop_n(ctx, nrets - 1);
+	}
```

Comment:
This seems a bit awkward (I guess it's related to the safe call issue). Couldn't the resolve function simply be trusted to leave the stack in a certain shape (i.e. never return > 1). Or maybe throw an error for `nrets > 1`?

The popping behavior seems a bit odd because it doesn't actually match Duktape/C behavior because the topmost value is discarded.